File: 102.binary_tree_level_order_traversal/levelOrder.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrderBottom = function(root) {
    var result = [];
    if(root === null) return result;

    var curLevelCount = 1;
    var nextLevelCount = 0;
    var temp =[];//value array
    var queue = [];//node array
    queue.push(root);
    while(queue.length !== 0){
        curLevelCount--;
        var node = queue.shift();
        temp.push(node.val);
        if(node.left !== null){
            queue.push(node.left);
            nextLevelCount++;
        }
        if(node.right !== null){
            queue.push(node.right);
            nextLevelCount++;
        }
        if(curLevelCount === 0){
            result.unshift(temp);
            curLevelCount = nextLevelCount;
            nextLevelCount = 0;
            temp = [];
        }

    }
    return result;
};

File: 107.binary_tree_level_order_traversal2/levelOrder.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from collections import deque


class Solution:
    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        queue = deque()
        queue.append(root)
        res = []
        while queue:
            nextsize = len(queue)
            level = []
            while nextsize > 0:
                node = queue.popleft()
                level.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
                nextsize -= 1
            res.insert(0, level)
